snapcraft/commands/core22/upload.py

```diff
@@ -18,20 +18,46 @@
 
 import pathlib
 import textwrap
-from typing import TYPE_CHECKING, List, Optional
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Any, List, Optional
 
 from craft_cli import BaseCommand, emit
 from craft_cli.errors import ArgumentParsingError
 from overrides import overrides
 from requests_toolbelt import MultipartEncoder, MultipartEncoderMonitor
 
-from snapcraft import store, utils
+from snapcraft import errors, store, utils
 from snapcraft_legacy._store import get_data_from_snap_file
 
 if TYPE_CHECKING:
     import argparse
 
 
+@dataclass(frozen=True)
+class ComponentOption:
+    """Argparse helper to validate and convert a 'component' option.
+
+    Receives a callable to convert the string from command line to the desired object.
+
+    Example of use:
+
+        parser.add_argument('--component',  type=ComponentOption())
+    """
+
+    name: str | None = None
+    filename: str | None = None
+
+    def __call__(self, value):
+        """Run by argparse to validate and convert the given argument."""
+        parts = [x.strip() for x in value.split("=")]
+        parts = [p for p in parts if p]
+        if len(parts) == 2:
```

Comment:
Another way to implement this would be `parts = tuple(map(str.strip, value.split("=", 1)))`.

Reply:
It's a good idea, but it doesn't work for some scenarios like a missing component name (`--component "=filename"`).

I did change it to be a one-liner and it should be cleaner.